### PR TITLE
Add PIO namelist control for CICE

### PR DESCRIPTION
**Pull request by DeniseWorthen**
## Commit Queue Requirements:

- [x] Fill out all sections of this template.
- [x] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [x] Commit 'test_changes.list' from previous step
---
## Description:

Adds namelist configuration of PIO options for CICE and switches from using the netcdf (serial) ``CICE_IO`` API to using PIO. This means all netCDF goes through a different API in CICE (ie, under ``infrastructure``, the code in ``io_pio2`` is built and used instead of the code in ``io_netcdf``). Also adds netcdf and PIO error checking. 

The available IO types are:

```
PIO_IOTYPE_PNETCDF=1 Parallel Netcdf (parallel)
PIO_IOTYPE_NETCDF=2 Netcdf3 Classic format (serial)
PIO_IOTYPE_NETCDF4C=3 NetCDF4 (HDF5) compressed format (serial)
PIO_IOTYPE_NETCDF4P=4 NetCDF4 (HDF5) parallel
```

Independent PIO settings are used for restart and history files.  New export variables are added to ``default_vars`` to allow all available settings to be used. Settings which are not relevant to a given PIO IOtype are ignored. If non-valid settings are given, appropriate values will be substituted. 

The available settings are

**(history,restart)_format**:
- Options are ``cdf1,2,5``, ``hdf5`` (netcdf4), ``pnetcdf1,2,5``. For cdf and pnetcdf, the 1,2,5 refer to netcdf3-classic, netcdf3-64-bit-offset and netcdf3-64bit-data. 

**(history,restart)_iotasks**:
- The subset of compute tasks which should also be used for IO. If not provided, will be given as the 1/4 of the available compute tasks. 

**(history,restart)_stride**:
- The stride of IO tasks across available compute tasks. If not given, will be set at 4. 

**(history,restart)_rearranger**: 
- For pnetcdf, either box or subset. Box is the default, but subset scales better. For small (< ~100 compute tasks), testing shows little difference between box and rearranger. Subset shows clear advantages at high processor numbers.  See [CICE IO tests](https://docs.google.com/spreadsheets/d/1xD0-gvbfI2Nwhf-ys_JdQEHR4Wibb0U5hGyVBBEqNUg/edit#gid=1262740475). 

**(history,restart)_root**:
- The processor used for the stride start

**(history,restart)_chunksize**, **(history,restart)_deflate**:
- For hdf5, the chunking and compression level

## Testing ##

The switch to PIO was not expected to change baselines compared to a baseline generated using the netCDF API. However, because of issues with the IC files, change were found in some, but not all tests using CICE. 

###  _**All tests using CICE should therefore get new baselines to ensure they are now consistent w/ using the PIO interface.**_ ###

 A test was run to verify that a baseline generated using PIO did reproduce itself. All tests using CICE passed.

### Commit Message:

```
Update to CICE-Consortium/CICE aca8357. Adds implementation of namelist PIO options for CICE
```

### Priority:

* Critical Bugfix: Reason
* High: Reason
* Normal

### UFSWM:

- related to #2163 

### Sub component Pull Requests:

* CICE: - https://github.com/NOAA-EMC/CICE/pull/74

### UFSWM Blocking Dependencies:

* None.

---
### Regression Test Changes (Please commit test_changes.list):

* PR Adds New Tests/Baselines.
* PR Updates/Changes Baselines.
* No Baseline Changes.

### Input data Changes:

* None.

### Library Changes/Upgrades:

* No Updates
  
---

## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] Jet
  - [x] Gaea
  - [x] Derecho
- WCOSS2
  - [x] Dogwood/Cactus
  - [x] Acorn
- [x] CI
- [ ] opnReqTest (complete task if unnecessary)

### CODE REVIEW

**Comment by DeniseWorthen** [comment]
@junwang-noaa This PR will *not* resolve the issue w/ the ice ICs since that will require some development work on the downscaling function, but it will enable PIO for CICE.

File: tests/parm/ice_in.IN [hunk 1]
```diff
@@ -1,48 +1,61 @@
...
-    histfreq_n     = @[CICE_HISTFREQ_N]
-    hist_avg       = @[CICE_HIST_AVG]
-    history_dir    = '@[CICE_HISTORY_DIR]'
-    history_file   = 'iceh'
-    write_ic       = .true.
-    incond_dir     = '@[CICE_INCOND_DIR]'
-    incond_file    = 'iceh_ic'
-    version_name   = 'CICE_6.0.2'
+    days_per_year	= 365
+    use_leap_years	= .true.
```
Comment (BrianCurtis-NOAA) [inline tests/parm/ice_in.IN]: These all have different spacing where the = sign is. At least on my phone. Can you double check?

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

File: tests/parm/ice_in.IN (hunk 1 above)
Comment (DeniseWorthen) [reply tests/parm/ice_in.IN]: I aligned the namelist on the ``=`` sign.

Comment (BrianCurtis-NOAA) [reply tests/parm/ice_in.IN]: Screenshot from 2024-04-05 08-04-32
This is how it looks on my laptop. Might want to verify your editor has the = sign aligned for the entire section.

Comment (DusanJovic-NOAA) [reply tests/parm/ice_in.IN]: Tabs vs spaces.

Comment (DeniseWorthen) [reply tests/parm/ice_in.IN]: @DusanJovic-NOAA Thanks. @BrianCurtis-NOAA Does that fix it?

**Review by BrianCurtis-NOAA — DISMISSED** [review]

File: CICE-interface/CMakeLists.txt [hunk 2]
```diff
@@ -39,7 +39,7 @@ else()
 endif()
 
 # Configuration Options
-set(CICE_IO "NetCDF" CACHE STRING "CICE OPTIONS: Choose IO options.")
+set(CICE_IO "PIO" CACHE STRING "CICE OPTIONS: Choose IO options.")
```
Comment (junwang-noaa) [inline CICE-interface/CMakeLists.txt:42]: Will this set PIO as default option for CICE?

**Review by junwang-noaa — DISMISSED** [review]

**Comment by FernandoAndrade-NOAA** [comment]
@zach1221 @BrianCurtis-NOAA FYI getting started on testing for this PR now.

**Comment by BrianCurtis-NOAA** [comment]
FYI WCOSS2 is unavailable this AM, but if all goes well will be back by this afternoon. So no need to rush this PR today.

**Comment by FernandoAndrade-NOAA** [comment]
Leaving a note here that on Hera, the `regional_atmaq_debug` intel timed out during baseline recreation requiring another 5 minute increase to the limit. On the matching run, the `cpld_debug_pdlib_p8_gnu` timed out as well, but this may have been due to a connection issue, I'm rerunning without a time increase first to verify.

```
69  47: WARNING: Open MPI failed to TCP connect to a peer MPI process.  This
     70  47: should not happen.
     71  47:
     72  47: Your Open MPI job may now hang or fail.
     73  47:
     74  47:   Local host: h25c19
     75  47:   PID:        652092
     76  47:   Message:    connect() to 10.184.25.25:1026 failed
     77  47:   Error:      Resource temporarily unavailable (11)
     78  47: --------------------------------------------------------------------------
     79  32: [h15c46][[45553,0],32][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(161) failed: Connection reset by peer (104)
     80  47: [h25c19:652092] 5 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
     81  47: [h25c19:652092] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
     82   0: slurmstepd: error: *** STEP 58110449.0 ON h15c46 CANCELLED AT 2024-04-09T10:32:42 DUE TO TIME LIMIT ***
     83 slurmstepd: error: *** JOB 58110449 ON h15c46 CANCELLED AT 2024-04-09T10:32:42 DUE TO TIME LIMIT ***
```

**Comment by BrianCurtis-NOAA** [comment]
@FernandoAndrade-NOAA I think if there was a timeout and this state that the job may hang/fail it probably did.

I think this should be brought up to the Hera SA's.

**Comment by FernandoAndrade-NOAA** [comment]
> @FernandoAndrade-NOAA I think if there was a timeout and this state that the job may hang/fail it probably did.
> 
> I think this should be brought up to the Hera SA's.

I'll go ahead and let them know, same error with more connection failures on the reruns and eventual timeout:
```
   Error:      Resource temporarily unavailable (11)
    101 220: --------------------------------------------------------------------------
    102 203: [h2c54:283633] 2 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
    103 203: [h2c54:283633] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
    104 184: [h2c34][[52227,0],184][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(248) failed: Connection reset by peer (104)
    105 193: [h2c34][[52227,0],193][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(180) failed: Connection reset by peer (104)
    106 195: [h2c34][[52227,0],195][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(256) failed: Connection reset by peer (104)
```

**Comment by jkbk2004** [comment]
> > @FernandoAndrade-NOAA I think if there was a timeout and this state that the job may hang/fail it probably did.
> > I think this should be brought up to the Hera SA's.
> 
> I'll go ahead and let them know, same error with more connection failures on the reruns and eventual timeout:
> 
> ```
>    Error:      Resource temporarily unavailable (11)
>     101 220: --------------------------------------------------------------------------
>     102 203: [h2c54:283633] 2 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
>     103 203: [h2c54:283633] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
>     104 184: [h2c34][[52227,0],184][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(248) failed: Connection reset by peer (104)
>     105 193: [h2c34][[52227,0],193][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(180) failed: Connection reset by peer (104)
>     106 195: [h2c34][[52227,0],195][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(256) failed: Connection reset by peer (104)
> ```

@RatkoVasic-NOAA @ulmononian this is another troubleshooting case related with pdlib side. we will double check with develop and hercules side. will keep you posted.

**Comment by FernandoAndrade-NOAA** [comment]
> > > @FernandoAndrade-NOAA I think if there was a timeout and this state that the job may hang/fail it probably did.
> > > I think this should be brought up to the Hera SA's.
> > 
> > 
> > I'll go ahead and let them know, same error with more connection failures on the reruns and eventual timeout:
> > ```
> >    Error:      Resource temporarily unavailable (11)
> >     101 220: --------------------------------------------------------------------------
> >     102 203: [h2c54:283633] 2 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
> >     103 203: [h2c54:283633] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
> >     104 184: [h2c34][[52227,0],184][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(248) failed: Connection reset by peer (104)
> >     105 193: [h2c34][[52227,0],193][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(180) failed: Connection reset by peer (104)
> >     106 195: [h2c34][[52227,0],195][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(256) failed: Connection reset by peer (104)
> > ```
> 
> @RatkoVasic-NOAA @ulmononian this is another troubleshooting case related with pdlib side. we will double check with develop and hercules side. will keep you posted.

Testing against develop and again with spack-stack 1.6.0 on hercules both resulted in segmentation faults.
develop: `/work2/noaa/epic/nandoam/stmp/nandoam/FV3_RT/rt_1379422/cpld_debug_pdlib_p8_gnu/err`
develop with spack stack 1.6.0: `/work2/noaa/epic/nandoam/stmp/nandoam/FV3_RT/rt_1366408/cpld_debug_pdlib_p8_gnu/err`

**Comment by jkbk2004** [comment]
> > > > @FernandoAndrade-NOAA I think if there was a timeout and this state that the job may hang/fail it probably did.
> > > > I think this should be brought up to the Hera SA's.
> > > 
> > > 
> > > I'll go ahead and let them know, same error with more connection failures on the reruns and eventual timeout:
> > > ```
> > >    Error:      Resource temporarily unavailable (11)
> > >     101 220: --------------------------------------------------------------------------
> > >     102 203: [h2c54:283633] 2 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
> > >     103 203: [h2c54:283633] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
> > >     104 184: [h2c34][[52227,0],184][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(248) failed: Connection reset by peer (104)
> > >     105 193: [h2c34][[52227,0],193][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(180) failed: Connection reset by peer (104)
> > >     106 195: [h2c34][[52227,0],195][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(256) failed: Connection reset by peer (104)
> > > ```
> > 
> > 
> > @RatkoVasic-NOAA @ulmononian this is another troubleshooting case related with pdlib side. we will double check with develop and hercules side. will keep you posted.
> 
> Testing against develop and again with spack-stack 1.6.0 on hercules both resulted in segmentation faults. develop: `/work2/noaa/epic/nandoam/stmp/nandoam/FV3_RT/rt_1379422/cpld_debug_pdlib_p8_gnu/err` develop with spack stack 1.6.0: `/work2/noaa/epic/nandoam/stmp/nandoam/FV3_RT/rt_1366408/cpld_debug_pdlib_p8_gnu/err`

Crashing points to the line https://github.com/NOAA-EMC/MOM6/blob/ab7bd14d209592d55490e75dbfaa61cb4a62df97/config_src/infra/FMS2/MOM_io_infra.F90#L905

**Comment by jkbk2004** [comment]
/scratch1/NCEPDEV/stmp2/Jong.Kim/FV3_RT/rt_1375704/cpld_debug_pdlib_p8_gnu runs ok on hera with develop branch. @FernandoAndrade-NOAA can you rerun the case with pr?

**Comment by DeniseWorthen** [comment]
I successfully ran the pdlib debug gnu test on Hera on the 4th. See 
```
/scratch1/NCEPDEV/stmp2/Denise.Worthen/FV3_RT/rt_1718506/cpld_debug_pdlib_p8_gnu
```

Has something changed on the system since then?

**Comment by DeniseWorthen** [comment]
Today, I get the same failure reported above, but I'll note it seems to be coming from the WAV PETS (180-257). 

```
221: [h6c35:883934] 6 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
221: [h6c35:883934] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
228: [h6c35:883941] 5 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
228: [h6c35:883941] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
237: [h6c35:883950] 5 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
237: [h6c35:883950] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
203: [h6c35:883916] 3 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
203: [h6c35:883916] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
218: [h6c35:883931] 1 more process has sent help message help-mpi-btl-tcp.txt / client connect fail
218: [h6c35:883931] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
220: [h6c35:883933] 3 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
220: [h6c35:883933] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
222: [h6c35:883935] 9 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
222: [h6c35:883935] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
226: [h6c35:883939] 1 more process has sent help message help-mpi-btl-tcp.txt / client connect fail
226: [h6c35:883939] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
238: [h6c35:883951] 7 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
238: [h6c35:883951] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

**Comment by DeniseWorthen** [comment]
I'm testing w/ a higher task count for Waves and it is currently running. I'm getting the sense that R8 is more sensitive to memory use. I saw it in a recent ufs-utils PR that I have been testing.

**Comment by DeniseWorthen** [comment]
The following change allows the cpld_debug_pdlib_gnu test to complete on Hera/R8/GNU

```
diff --git a/tests/tests/cpld_debug_pdlib_p8 b/tests/tests/cpld_debug_pdlib_p8
index fa7be3c0..45cb9c15 100644
--- a/tests/tests/cpld_debug_pdlib_p8
+++ b/tests/tests/cpld_debug_pdlib_p8
@@ -73,7 +73,7 @@ OCN_tasks=$OCN_tasks_cpl_unstr
 ICE_tasks=$ICE_tasks_cpl_unstr
 WAV_tasks=$WAV_tasks_cpl_unstr
 # bump resources for debug test
-WAV_tasks="$(($WAV_tasks_cpl_unstr + 18))"
+WAV_tasks="$(($WAV_tasks_cpl_unstr + 40))"
```

The times reported in the log were

```
PASS -- COMPILE 's2sw_pdlib_debug_gnu' [04:07, 02:28]
PASS -- TEST 'cpld_debug_pdlib_p8_gnu' [18:04, 11:52](1311 MB)
```

**Comment by FernandoAndrade-NOAA** [comment]
> /scratch1/NCEPDEV/stmp2/Jong.Kim/FV3_RT/rt_1375704/cpld_debug_pdlib_p8_gnu runs ok on hera with develop branch. @FernandoAndrade-NOAA can you rerun the case with pr?

It's passing for me now, I'm not sure what may have been the cause of this issue. I'll push hera up shortly.

**Comment by FernandoAndrade-NOAA** [comment]
Ok, all logs have been pushed up, we can continue with the merge process in CICE.

**Comment by DeniseWorthen** [comment]
@BrianCurtis-NOAA Would you please create an issue for removing the WCOSS2 setting in default_vars, once we know PIO is correctly working?

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Review by jkbk2004 — APPROVED** [review]

**Pull request merged**